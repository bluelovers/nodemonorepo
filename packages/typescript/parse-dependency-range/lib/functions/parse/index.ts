import url from 'url'
import semver from 'semver'
import { array } from 'convenient-typescript-utilities'
import { Struct, Type, GitUrl, LocalUrl, TarballUrl } from '../../types'
import { GITHUB_SHORTHAND, GIT_URL, LOCAL_URL, TARBALL_URL } from '../../constants'
const { includes } = array

function parse (value: string): Struct {
  if (value === 'latest') {
    return {
      type: Type.Latest,
      value
    }
  }

  const range: string | null = semver.validRange(value)
  if (range) {
    return {
      type: Type.Semver,
      value,
      range
    }
  }

  if (GITHUB_SHORTHAND.test(value)) {
    const [org, ...orgRest] = value.split('/')
    const [repo, ...repoRest] = orgRest.join('/').split('#')
    const hash = repoRest.join('#') || undefined

    return {
      type: Type.GitHub,
      value,
      org,
      repo,
      hash
    }
  }

  const urlObject = url.parse(value)
  const { protocol } = urlObject

  if (includes(GIT_URL.PROTOCOL, protocol)) {
    return {
      type: Type.Git,
      value,
      url: urlObject as GitUrl
    }
  }

  if (includes(LOCAL_URL.PROTOCOL, protocol)) {
    return {
      type: Type.Local,
      value,
      path: value.split(':').slice(1).join(':'),
      protocol: protocol as LocalUrl.Protocol
    }
  }

  if (includes(TARBALL_URL.PROTOCOL, protocol)) {
    const extension: TarballUrl.Extension | undefined =
      TARBALL_URL.EXTENSION.find(ext => value.endsWith(ext))

    if (extension) {
      return {
        type: Type.Tarball,
        value,
        url: {
          ...urlObject,
          extension
        } as TarballUrl
      }
    }
  }

  return {
    type: Type.Unknown,
    value
  }
}

export = parse
