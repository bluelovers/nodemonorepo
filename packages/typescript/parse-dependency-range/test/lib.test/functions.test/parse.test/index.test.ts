import parse, { Type } from '../../../../index'

type Sample<X extends Type = Type> = [X, ReadonlyArray<string>]

const samples: ReadonlyArray<Sample> = [
  [Type.Semver, [
    '0.1.2',
    '=0.1.2',
    '~0.1.2',
    '^0.1.2',
    '*'
  ]],
  [Type.Tarball, [
    'ftp://example.com/example.tgz',
    'ftps://example.com/example.tgz',
    'ftp://example.com/example.tar.gzip',
    'ftps://example.com/example.tar.gzip',
    'http://example.com/example.tgz',
    'https://example.com/example.tgz',
    'http://example.com/example.tar.gzip',
    'https://example.com/example.tar.gzip'
  ]],
  [Type.Git, [
    'git://example.com/org/repo.git#v0.1.2',
    'git+ssh://git@example.com:org/repo.git#v0.1.2',
    'git+ssh://git@example.com:org/repo#semver:^5.0',
    'git+http://user@example.com/org/repo.git',
    'git+https://user@example.com/org/repo.git',
    'git+file:../../hello/world'
  ]],
  [Type.GitHub, [
    'org/repo',
    'org/repo#hash',
    'org/repo#hash/tag',
    'org-name/repo-name',
    'org-name/repo-name#commit-hash',
    'org-name/repo-name#commit-hash/tag-name',
    'org_name/repo_name',
    'org_name/repo_name#commit_hash',
    'org_name/repo_name#commit_hash/tag_name',
    'org.name/repo.name',
    'org.name/repo.name#commit.hash',
    'org.name/repo.name#commit.hash/tag.name',
    'org123/repo456',
    'org123/repo456#hash789',
    'org123/repo456#hash789/tag',
    'Org/Repo',
    'Org/Repo#Hash',
    'Org/Repo#Hash/Tag',
    'ORG/REPO',
    'ORG/REPO#HASH',
    'ORG/REPO#HASH/TAG'
  ]],
  [Type.Local, [
    'file:///home/user/package',
    'file:../../hello/world,',
    'link:///home/user/package',
    'link:../../hello/world,'
  ]],
  [Type.Latest, [
    'latest'
  ]],
  [Type.Unknown, [
    '!@#$%^&*()',
    'blahblah'
  ]]
]

describe('when using samples', () => {
  const results = samples.map(([type, array]) => {
    const table = array.map(input => ({
      input,
      output: parse(input)
    }))

    return { type, table }
  })

  it('emits correct types', () => {
    const received = results.map(({ table }) => table.map(({ input, output }) => ({ input, type: output.type })))
    const expected = results.map(({ table, type }) => table.map(({ input }) => ({ input, type })))
    expect(received).toEqual(expected)
  })

  it('emits correct values', () => {
    const received = results.map(x => x.table.map(({ input, output }) => ({ input, value: output.value })))
    const expected = results.map(x => x.table.map(({ input }) => ({ input, value: input })))
    expect(received).toEqual(expected)
  })

  it('matches snapshot', () => {
    expect(results).toMatchSnapshot()
  })
})
