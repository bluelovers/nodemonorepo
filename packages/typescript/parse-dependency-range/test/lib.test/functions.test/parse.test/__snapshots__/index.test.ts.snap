// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when using samples matches snapshot 1`] = `
Array [
  Object {
    "table": Array [
      Object {
        "input": "0.1.2",
        "output": Object {
          "range": "0.1.2",
          "type": "semver",
          "value": "0.1.2",
        },
      },
      Object {
        "input": "=0.1.2",
        "output": Object {
          "range": "0.1.2",
          "type": "semver",
          "value": "=0.1.2",
        },
      },
      Object {
        "input": "~0.1.2",
        "output": Object {
          "range": ">=0.1.2 <0.2.0",
          "type": "semver",
          "value": "~0.1.2",
        },
      },
      Object {
        "input": "^0.1.2",
        "output": Object {
          "range": ">=0.1.2 <0.2.0",
          "type": "semver",
          "value": "^0.1.2",
        },
      },
      Object {
        "input": "*",
        "output": Object {
          "range": "*",
          "type": "semver",
          "value": "*",
        },
      },
    ],
    "type": "semver",
  },
  Object {
    "table": Array [
      Object {
        "input": "ftp://example.com/example.tgz",
        "output": Object {
          "type": "tarball",
          "url": Object {
            "auth": null,
            "extension": ".tgz",
            "hash": null,
            "host": "example.com",
            "hostname": "example.com",
            "href": "ftp://example.com/example.tgz",
            "path": "/example.tgz",
            "pathname": "/example.tgz",
            "port": null,
            "protocol": "ftp:",
            "query": null,
            "search": null,
            "slashes": true,
          },
          "value": "ftp://example.com/example.tgz",
        },
      },
      Object {
        "input": "ftps://example.com/example.tgz",
        "output": Object {
          "type": "tarball",
          "url": Object {
            "auth": null,
            "extension": ".tgz",
            "hash": null,
            "host": "example.com",
            "hostname": "example.com",
            "href": "ftps://example.com/example.tgz",
            "path": "/example.tgz",
            "pathname": "/example.tgz",
            "port": null,
            "protocol": "ftps:",
            "query": null,
            "search": null,
            "slashes": true,
          },
          "value": "ftps://example.com/example.tgz",
        },
      },
      Object {
        "input": "ftp://example.com/example.tar.gzip",
        "output": Object {
          "type": "tarball",
          "url": Object {
            "auth": null,
            "extension": ".tar.gzip",
            "hash": null,
            "host": "example.com",
            "hostname": "example.com",
            "href": "ftp://example.com/example.tar.gzip",
            "path": "/example.tar.gzip",
            "pathname": "/example.tar.gzip",
            "port": null,
            "protocol": "ftp:",
            "query": null,
            "search": null,
            "slashes": true,
          },
          "value": "ftp://example.com/example.tar.gzip",
        },
      },
      Object {
        "input": "ftps://example.com/example.tar.gzip",
        "output": Object {
          "type": "tarball",
          "url": Object {
            "auth": null,
            "extension": ".tar.gzip",
            "hash": null,
            "host": "example.com",
            "hostname": "example.com",
            "href": "ftps://example.com/example.tar.gzip",
            "path": "/example.tar.gzip",
            "pathname": "/example.tar.gzip",
            "port": null,
            "protocol": "ftps:",
            "query": null,
            "search": null,
            "slashes": true,
          },
          "value": "ftps://example.com/example.tar.gzip",
        },
      },
      Object {
        "input": "http://example.com/example.tgz",
        "output": Object {
          "type": "tarball",
          "url": Object {
            "auth": null,
            "extension": ".tgz",
            "hash": null,
            "host": "example.com",
            "hostname": "example.com",
            "href": "http://example.com/example.tgz",
            "path": "/example.tgz",
            "pathname": "/example.tgz",
            "port": null,
            "protocol": "http:",
            "query": null,
            "search": null,
            "slashes": true,
          },
          "value": "http://example.com/example.tgz",
        },
      },
      Object {
        "input": "https://example.com/example.tgz",
        "output": Object {
          "type": "tarball",
          "url": Object {
            "auth": null,
            "extension": ".tgz",
            "hash": null,
            "host": "example.com",
            "hostname": "example.com",
            "href": "https://example.com/example.tgz",
            "path": "/example.tgz",
            "pathname": "/example.tgz",
            "port": null,
            "protocol": "https:",
            "query": null,
            "search": null,
            "slashes": true,
          },
          "value": "https://example.com/example.tgz",
        },
      },
      Object {
        "input": "http://example.com/example.tar.gzip",
        "output": Object {
          "type": "tarball",
          "url": Object {
            "auth": null,
            "extension": ".tar.gzip",
            "hash": null,
            "host": "example.com",
            "hostname": "example.com",
            "href": "http://example.com/example.tar.gzip",
            "path": "/example.tar.gzip",
            "pathname": "/example.tar.gzip",
            "port": null,
            "protocol": "http:",
            "query": null,
            "search": null,
            "slashes": true,
          },
          "value": "http://example.com/example.tar.gzip",
        },
      },
      Object {
        "input": "https://example.com/example.tar.gzip",
        "output": Object {
          "type": "tarball",
          "url": Object {
            "auth": null,
            "extension": ".tar.gzip",
            "hash": null,
            "host": "example.com",
            "hostname": "example.com",
            "href": "https://example.com/example.tar.gzip",
            "path": "/example.tar.gzip",
            "pathname": "/example.tar.gzip",
            "port": null,
            "protocol": "https:",
            "query": null,
            "search": null,
            "slashes": true,
          },
          "value": "https://example.com/example.tar.gzip",
        },
      },
    ],
    "type": "tarball",
  },
  Object {
    "table": Array [
      Object {
        "input": "git://example.com/org/repo.git#v0.1.2",
        "output": Object {
          "type": "git",
          "url": Url {
            "auth": null,
            "hash": "#v0.1.2",
            "host": "example.com",
            "hostname": "example.com",
            "href": "git://example.com/org/repo.git#v0.1.2",
            "path": "/org/repo.git",
            "pathname": "/org/repo.git",
            "port": null,
            "protocol": "git:",
            "query": null,
            "search": null,
            "slashes": true,
          },
          "value": "git://example.com/org/repo.git#v0.1.2",
        },
      },
      Object {
        "input": "git+ssh://git@example.com:org/repo.git#v0.1.2",
        "output": Object {
          "type": "git",
          "url": Url {
            "auth": "git",
            "hash": "#v0.1.2",
            "host": "example.com",
            "hostname": "example.com",
            "href": "git+ssh://git@example.com/:org/repo.git#v0.1.2",
            "path": "/:org/repo.git",
            "pathname": "/:org/repo.git",
            "port": null,
            "protocol": "git+ssh:",
            "query": null,
            "search": null,
            "slashes": true,
          },
          "value": "git+ssh://git@example.com:org/repo.git#v0.1.2",
        },
      },
      Object {
        "input": "git+ssh://git@example.com:org/repo#semver:^5.0",
        "output": Object {
          "type": "git",
          "url": Url {
            "auth": "git",
            "hash": "#semver:%5E5.0",
            "host": "example.com",
            "hostname": "example.com",
            "href": "git+ssh://git@example.com/:org/repo#semver:%5E5.0",
            "path": "/:org/repo",
            "pathname": "/:org/repo",
            "port": null,
            "protocol": "git+ssh:",
            "query": null,
            "search": null,
            "slashes": true,
          },
          "value": "git+ssh://git@example.com:org/repo#semver:^5.0",
        },
      },
      Object {
        "input": "git+http://user@example.com/org/repo.git",
        "output": Object {
          "type": "git",
          "url": Url {
            "auth": "user",
            "hash": null,
            "host": "example.com",
            "hostname": "example.com",
            "href": "git+http://user@example.com/org/repo.git",
            "path": "/org/repo.git",
            "pathname": "/org/repo.git",
            "port": null,
            "protocol": "git+http:",
            "query": null,
            "search": null,
            "slashes": true,
          },
          "value": "git+http://user@example.com/org/repo.git",
        },
      },
      Object {
        "input": "git+https://user@example.com/org/repo.git",
        "output": Object {
          "type": "git",
          "url": Url {
            "auth": "user",
            "hash": null,
            "host": "example.com",
            "hostname": "example.com",
            "href": "git+https://user@example.com/org/repo.git",
            "path": "/org/repo.git",
            "pathname": "/org/repo.git",
            "port": null,
            "protocol": "git+https:",
            "query": null,
            "search": null,
            "slashes": true,
          },
          "value": "git+https://user@example.com/org/repo.git",
        },
      },
      Object {
        "input": "git+file:../../hello/world",
        "output": Object {
          "type": "git",
          "url": Url {
            "auth": null,
            "hash": null,
            "host": "..",
            "hostname": "..",
            "href": "git+file:../../hello/world",
            "path": "/../hello/world",
            "pathname": "/../hello/world",
            "port": null,
            "protocol": "git+file:",
            "query": null,
            "search": null,
            "slashes": null,
          },
          "value": "git+file:../../hello/world",
        },
      },
    ],
    "type": "git",
  },
  Object {
    "table": Array [
      Object {
        "input": "org/repo",
        "output": Object {
          "hash": undefined,
          "org": "org",
          "repo": "repo",
          "type": "github",
          "value": "org/repo",
        },
      },
      Object {
        "input": "org/repo#hash",
        "output": Object {
          "hash": "hash",
          "org": "org",
          "repo": "repo",
          "type": "github",
          "value": "org/repo#hash",
        },
      },
      Object {
        "input": "org/repo#hash/tag",
        "output": Object {
          "hash": "hash/tag",
          "org": "org",
          "repo": "repo",
          "type": "github",
          "value": "org/repo#hash/tag",
        },
      },
      Object {
        "input": "org-name/repo-name",
        "output": Object {
          "hash": undefined,
          "org": "org-name",
          "repo": "repo-name",
          "type": "github",
          "value": "org-name/repo-name",
        },
      },
      Object {
        "input": "org-name/repo-name#commit-hash",
        "output": Object {
          "hash": "commit-hash",
          "org": "org-name",
          "repo": "repo-name",
          "type": "github",
          "value": "org-name/repo-name#commit-hash",
        },
      },
      Object {
        "input": "org-name/repo-name#commit-hash/tag-name",
        "output": Object {
          "hash": "commit-hash/tag-name",
          "org": "org-name",
          "repo": "repo-name",
          "type": "github",
          "value": "org-name/repo-name#commit-hash/tag-name",
        },
      },
      Object {
        "input": "org_name/repo_name",
        "output": Object {
          "hash": undefined,
          "org": "org_name",
          "repo": "repo_name",
          "type": "github",
          "value": "org_name/repo_name",
        },
      },
      Object {
        "input": "org_name/repo_name#commit_hash",
        "output": Object {
          "hash": "commit_hash",
          "org": "org_name",
          "repo": "repo_name",
          "type": "github",
          "value": "org_name/repo_name#commit_hash",
        },
      },
      Object {
        "input": "org_name/repo_name#commit_hash/tag_name",
        "output": Object {
          "hash": "commit_hash/tag_name",
          "org": "org_name",
          "repo": "repo_name",
          "type": "github",
          "value": "org_name/repo_name#commit_hash/tag_name",
        },
      },
      Object {
        "input": "org.name/repo.name",
        "output": Object {
          "hash": undefined,
          "org": "org.name",
          "repo": "repo.name",
          "type": "github",
          "value": "org.name/repo.name",
        },
      },
      Object {
        "input": "org.name/repo.name#commit.hash",
        "output": Object {
          "hash": "commit.hash",
          "org": "org.name",
          "repo": "repo.name",
          "type": "github",
          "value": "org.name/repo.name#commit.hash",
        },
      },
      Object {
        "input": "org.name/repo.name#commit.hash/tag.name",
        "output": Object {
          "hash": "commit.hash/tag.name",
          "org": "org.name",
          "repo": "repo.name",
          "type": "github",
          "value": "org.name/repo.name#commit.hash/tag.name",
        },
      },
      Object {
        "input": "org123/repo456",
        "output": Object {
          "hash": undefined,
          "org": "org123",
          "repo": "repo456",
          "type": "github",
          "value": "org123/repo456",
        },
      },
      Object {
        "input": "org123/repo456#hash789",
        "output": Object {
          "hash": "hash789",
          "org": "org123",
          "repo": "repo456",
          "type": "github",
          "value": "org123/repo456#hash789",
        },
      },
      Object {
        "input": "org123/repo456#hash789/tag",
        "output": Object {
          "hash": "hash789/tag",
          "org": "org123",
          "repo": "repo456",
          "type": "github",
          "value": "org123/repo456#hash789/tag",
        },
      },
      Object {
        "input": "Org/Repo",
        "output": Object {
          "hash": undefined,
          "org": "Org",
          "repo": "Repo",
          "type": "github",
          "value": "Org/Repo",
        },
      },
      Object {
        "input": "Org/Repo#Hash",
        "output": Object {
          "hash": "Hash",
          "org": "Org",
          "repo": "Repo",
          "type": "github",
          "value": "Org/Repo#Hash",
        },
      },
      Object {
        "input": "Org/Repo#Hash/Tag",
        "output": Object {
          "hash": "Hash/Tag",
          "org": "Org",
          "repo": "Repo",
          "type": "github",
          "value": "Org/Repo#Hash/Tag",
        },
      },
      Object {
        "input": "ORG/REPO",
        "output": Object {
          "hash": undefined,
          "org": "ORG",
          "repo": "REPO",
          "type": "github",
          "value": "ORG/REPO",
        },
      },
      Object {
        "input": "ORG/REPO#HASH",
        "output": Object {
          "hash": "HASH",
          "org": "ORG",
          "repo": "REPO",
          "type": "github",
          "value": "ORG/REPO#HASH",
        },
      },
      Object {
        "input": "ORG/REPO#HASH/TAG",
        "output": Object {
          "hash": "HASH/TAG",
          "org": "ORG",
          "repo": "REPO",
          "type": "github",
          "value": "ORG/REPO#HASH/TAG",
        },
      },
    ],
    "type": "github",
  },
  Object {
    "table": Array [
      Object {
        "input": "file:///home/user/package",
        "output": Object {
          "path": "///home/user/package",
          "protocol": "file:",
          "type": "local",
          "value": "file:///home/user/package",
        },
      },
      Object {
        "input": "file:../../hello/world,",
        "output": Object {
          "path": "../../hello/world,",
          "protocol": "file:",
          "type": "local",
          "value": "file:../../hello/world,",
        },
      },
      Object {
        "input": "link:///home/user/package",
        "output": Object {
          "path": "///home/user/package",
          "protocol": "link:",
          "type": "local",
          "value": "link:///home/user/package",
        },
      },
      Object {
        "input": "link:../../hello/world,",
        "output": Object {
          "path": "../../hello/world,",
          "protocol": "link:",
          "type": "local",
          "value": "link:../../hello/world,",
        },
      },
    ],
    "type": "local",
  },
  Object {
    "table": Array [
      Object {
        "input": "latest",
        "output": Object {
          "type": "latest",
          "value": "latest",
        },
      },
    ],
    "type": "latest",
  },
  Object {
    "table": Array [
      Object {
        "input": "!@#$%^&*()",
        "output": Object {
          "type": "unknown",
          "value": "!@#$%^&*()",
        },
      },
      Object {
        "input": "blahblah",
        "output": Object {
          "type": "unknown",
          "value": "blahblah",
        },
      },
    ],
    "type": "unknown",
  },
]
`;
